Allow disabling alerts

Gudahtt opened the pull request:
The unconnected account alert can now be disabled. A "don't show this again" checkbox has been added to the alert, which prevents that alert from being shown in the future.

An alert settings page has been added to the settings as well. This page allows the user to disable or enable any alert.

Gudahtt committed:
Allow disabling alerts

The unconnected account alert can now be disabled. A "don't show this
again" checkbox has been added to the alert, which prevents that alert
from being shown in the future.

An alert settings page has been added to the settings as well. This
page allows the user to disable or enable any alert.

Gudahtt commented on ui/app/components/app/alerts/unconnected-account-alert/unconnected-account-alert.js:
```diff
@@ -56,7 +67,27 @@ const SwitchToUnconnectedAccountAlert = () => {
-    />
+    >
+      <Checkbox
+        id="unconnectedAccount_dontShowThisAgain"
```
This prop doesn't yet exist in `CheckBox`. I should make it exist.

Gudahtt replied:
This was done in #8553

Gudahtt replied:
I'm not going to consider this PR to be strictly dependant upon #8553 because this still works fine as-is, kinda. The prop is useless, and this is sub-optimal from an accessibility standpoint, but it's no worse than the rest of our extension. Plus #8553 will likely get merged soon anyway.

Gudahtt commented:
<details>
<summary>Screenshots:</summary>

"Unconnected account" alert with the new checkbox:

![alert-with-checkbox](https://user-images.githubusercontent.com/2459287/81358966-4be17f00-90ae-11ea-84e8-5ea35fff31a1.png)

"Unconnected account" alert with the new checkbox, showing the hover state of the info-icon:

![alerith-with-checkbox-and-hover](https://user-images.githubusercontent.com/2459287/81358986-54d25080-90ae-11ea-9ab2-59523cc68b42.png)

Alert settings tab in settings menu:

![alerts-settings-list](https://user-images.githubusercontent.com/2459287/81358772-ce1d7380-90ad-11ea-8cb3-88a9e58021c1.png)

Alert settings page:

![alerts-page](https://user-images.githubusercontent.com/2459287/81358792-d83f7200-90ad-11ea-8a89-fa02441c651d.png)

</details>

whymarrh commented on ui/app/pages/settings/alerts-tab/alerts-tab.js:
```diff
@@ -0,0 +1,70 @@
+  const t = useContext(I18nContext)
+
+  const alertConfig = {
+    unconnectedAccount: {
```
Can we put these keys in a single location? Currently this needs to match with the string we're using in the selector and that seems like it might be error-prone. Could we put the keys in the bg?

whymarrh replied:
Both of:

- `ui/app/ducks/metamask/metamask.js:373`
- `ui/app/ducks/alerts/unconnected-account.js:20`

Gudahtt replied:
Yeah, we can do that. I had considered it before - there was one revision where I exported the `name` variable in `ui/app/ducks/alerts/unconnected-account.js` and used that in a few places. It felt a bit tedious at the time, but now that it's in the background too it probably makes sense to add it as a constant somewhere.

Gudahtt replied:
This was done in 5c4df5538361d5238f4dc3100c47acc8f83e4db8

Gudahtt committed:
Add constant for alert type

The string `unconnectedAccount` is now stored as a constant. This was
done to make it more clear that this string had to match exactly in
each of these places.

whymarrh commented on app/scripts/controllers/alert.js:
```diff
@@ -0,0 +1,53 @@
+
+const defaultState = {
+  alertEnabledness: Object.keys(ALERT_TYPES)
+    .reduce(
+      (alertEnabledness, alertType) => {
+        alertEnabledness[alertType] = true
+        return alertEnabledness
+      },
+      {}
+    ),
```
The mutation of the accumulator here, while valid, feels weird. 😅 Maybe something like this:

```suggestion
    .reduce((alertEnabledness, alertType) => ({
      ...alertEnabledness,
      [alertType]: true,
    }), {}),
```

Gudahtt replied:
Huh, really? I mutated it intentionally. Why bother construct a new object each time?

whymarrh replied:
> Why bother construct a new object each time?

I won't disagree that mutation is faster, but the reduce construct is/feels like a functional construct that fits with a more "immutable" pattern. It's an unimportant nitpick.

Gudahtt replied:
Sure, fair enough. I still think of it as a"functional construct" because the accumulator is "internal" to this reduce step, so it's still effectively presenting an immutable interface. 🤷‍♂️

whymarrh reviewed (dismissed):
LGTM! (One small non-blocking suggestion.)

rachelcope reviewed (dismissed):
This looks good. Small copy suggestion on the tooltip > `This can be changed in "Settings > Alerts"`

Gudahtt committed:
Update copy of 'Don't show this again' tooltip

Gudahtt commented:
@rachelcope Thanks, that does look better. It has been updated here: e9bb1140cde97bd6ce88ce7678b2f7bdeae7a23a

whymarrh reviewed (dismissed):
LGTM!

Gudahtt commented on app/scripts/controllers/alert.js:
```diff
@@ -0,0 +1,53 @@
+
+  setAlertEnabledness (alertId, enabledness) {
+    const { alertEnabledness } = this.store.getState()
+    alertEnabledness[alertId] = enabledness
```
Speaking of mutation - I should not be mutating here 🤦‍♂️  I'll fix this.

Gudahtt committed:
Shallow clone state before updating to avoid mutation
